lib/active_fedora/rdf/indexing_service.rb

```diff
@@ -64,35 +64,107 @@ def resource
         object.resource
       end
 
-      def properties
-        object.class.properties
-      end
-
       def index_config
         object.class.index_config
       end
 
-      # returns a Hash, e.g.: {field => { values: [], type: :something, behaviors: [] }, ...}
+      # returns the field map instance
       def fields
-        field_map = {}.with_indifferent_access
-
-        index_config.each do |name, index_field_config|
-          type = index_field_config.data_type
-          behaviors = index_field_config.behaviors
-          next unless type and behaviors
-          next if kind_of_af_base?(name)
-          field_map[name] = { values: find_values(name), type: type, behaviors: behaviors}
+        field_map_class.new do |field_map| 
+          index_config.each { |name, index_field_config| field_map.insert(name, index_field_config, object) }
         end
-        field_map
       end
 
-      def kind_of_af_base?(name)
-        config = properties[name.to_s]
-        config && config[:class_name] && config[:class_name] < ActiveFedora::Base
+      # Override this method to use your own FieldMap class for custom indexing of objects and properties
+      def field_map_class
+        FieldMap
+      end
+
+      # Transient class that represents a field that we send to solr.
+      # It might be possible for two properties to share a single field map entry if they use the same solr key.
+      # @attribute [Symbol] type the data type hint for Solrizer
+      # @attribute [Array] behaviors the indexing hints such as :stored_searchable or :symbol
+      # @attribute [Array] values the actual values that get sent to solr
+      class FieldMapEntry
+
+        attr_accessor :type, :behaviors, :values
+
+        def initialize
+          @behaviors = []
+          @values = []
+        end
+
+        def merge!(type, behaviors, values)
+          self.type ||= type
+          self.behaviors += behaviors
+          self.behaviors.uniq!
+          self.values += values
+        end
+
       end
 
-      def find_values(name)
-        object.send(name) || []
+      # Transient class that maps solr field names, without their suffixes, to the values and behaviors that
+      # are used to transforming them for insertion into the solr document.
+      # It partially extends Ruby's Hash class, similar to the way ActiveFedora::Indexing::Map does,
+      # but only with selected methods as outlined in def_delegators.
+      class FieldMap
+        extend Forwardable
+        
+        def_delegators :@hash, :[], :[]=, :each, :keys
+        
+        def initialize(hash = {}, &block)
+          @hash = hash
+          yield self if block_given?
+        end
+
+        def insert(name, index_field_config, object)
+          self[index_field_config.key.to_s] ||= FieldMapEntry.new
+          builder = kind_of_af_base?(name, object) ? resource_builder_class : property_builder_class
```

Comment:
To clarify, I'm imagining something like this...
```ruby
class PolymorphicBuilder
  def build(entry, index_field_config, object, name)
    delegate_class(name, object).build(entry, index_field_config, object, name)
  end

  private

  def delegate_class(name, object)
    if kind_of_af_base?(name, object)
      ResourceBuilder
    else
      PropertyBuilder
    end
  end

  def kind_of_af_base(name, object)
    config = properties(object)[name.to_s]
    config && config[:class_name] && config[:class_name] < ActiveFedora::Base
  end

  def properties(object)
    object.class.properties
  end
end
```

When you write it out this way you start to see some patterns emerging too - object/name are passed around a lot, it seems like those may be something you'd want in a constructor.